chalk-solve/src/clauses/builtin_traits/unsize.rs

```diff
@@ -0,0 +1,449 @@
+}
+
+pub fn add_unsize_program_clauses<I: Interner>(
+    db: &dyn RustIrDatabase<I>,
+    builder: &mut ClauseBuilder<'_, I>,
+    trait_ref: &TraitRef<I>,
+    _ty: &TyData<I>,
+) {
+    let interner = db.interner();
+
+    let source_ty = trait_ref.self_type_parameter(interner);
+    let target_ty = trait_ref
+        .substitution
+        .at(interner, 1)
+        .assert_ty_ref(interner);
+
+    let unsize_trait_id = trait_ref.trait_id;
+
+    // N.B. here rustc asserts that `TraitRef` is not a higher-ranked bound
+    // i.e. `for<'a> &'a T: Unsize<dyn Trait+'a>` is never provable.
+    //
+    // In chalk it would be awkward to implement and I am not sure
+    // there is a need for it, the original comment states that this restriction
+    // could be lifted.
+    //
+    // for more info visit `fn assemble_candidates_for_unsizing` and
+    // `fn confirm_builtin_unisize_candidate` in rustc.
+
+    match (source_ty.data(interner), target_ty.data(interner)) {
+        // dyn Trait + AutoX + 'a -> dyn Trait + AutoY + 'b
+        (
+            TyData::Dyn(DynTy {
+                bounds: bounds_a,
+                lifetime: lifetime_a,
+            }),
+            TyData::Dyn(DynTy {
+                bounds: bounds_b,
+                lifetime: lifetime_b,
+            }),
+        ) => {
+            let principal_a = principal_id(db, bounds_a);
+            let principal_b = principal_id(db, bounds_b);
+
+            let auto_trait_ids_a: Vec<_> = auto_trait_ids(db, bounds_a).collect();
+            let auto_trait_ids_b: Vec<_> = auto_trait_ids(db, bounds_b).collect();
+
+            let may_apply = principal_a == principal_b
+                && auto_trait_ids_b
+                    .iter()
+                    .all(|id_b| auto_trait_ids_a.iter().any(|id_a| id_a == id_b));
+
+            if !may_apply {
+                return;
+            }
+
+            // COMMENT FROM RUSTC:
+            // Require that the traits involved in this upcast are **equal**;
+            // only the **lifetime bound** is changed.
+            //
+            // FIXME: This condition is arguably too strong -- it would
+            // suffice for the source trait to be a *subtype* of the target
+            // trait. In particular, changing from something like
+            // `for<'a, 'b> Foo<'a, 'b>` to `for<'a> Foo<'a, 'a>` should be
+            // permitted.
+            // <...>
+            // I've modified this to `.eq` because I want to continue rejecting
+            // that [`old-lub-glb-object.rs`] test (as we have
+            // done for quite some time) before we are firmly comfortable
+            // with what our behavior should be there. -nikomatsakis
+            // ------------------
+
+            // Filter out auto traits of source that are not present in target
+            // and change source lifetime to target lifetime
+            //
+            // This new type should be equal to target type.
```

Comment:
This isn't exactly true. This is what we *want*, but we don't know that.

Reply:
Agreed, this should be phrased better